Implement New GH Authentication Given Deprecation

lwasser opened the pull request:
this replaces 

https://github.com/earthlab/abc-classroom/pull/329  

closes #283 

i'm updating it with current  CI checks and plan to test locally. doing this so i can bump a new version  today with a suite of new fixes!

TODO's:

- [x] Fix linting - black and flake8
- [x] merge conflicts
- [x] update from main for current CI builds
- [x] fix docs
- [x] fix  changelog entries (added under an already released  version)
- [ ] TEST auth locally to ensure it does work

i  think all is working now but will see what CI has to say. 

**** ORIGINAL PR TEXT ***
This re-implements the abc-init script so that it no longer uses the soon-to-be-deprecated username + password authentication to generate a personal access token. We now have a GitHub App (abc-classroom-bot) and authenticate using the app to generate an OAuth access token. I still need to write docs, but this can be tested by running `abc-init`.

Addresses #328 and #326 
Also addresses #327


Note that the existing github3 calls work with the new access token, so replacing those does not need to be done immediately. Should still be done, but there isn't the same level of urgency as getting the token generation replaced.

lwasser commented:
failures right now  on  code coverage  we can ignore  for the time being

lwasser commented:
hey @kcranston   --  this  new authentication did NOT work for me. It did generate a token which  was ok but  wen i tried this
`abc-new-template my-assignment --github`

(i h ad  created a new test assignment) it did NOT push to  github. The error i got is below. i tried this  both at the CLI and as a package function and neither worked and both r eturned the same error

it's  returning a KeyError. Is there perhaps a change that i need to make to my config  file for this to work properly?

```
No changes to local repository.
Updating assignment list in config
Writing modified config at /Users/leahwasser/Documents/github/1-bootcamp-course/fall-2020-abc-bootcamp
Traceback (most recent call last):
  File "/Users/leahwasser/opt/miniconda3/envs/earth-analytics-python/bin/abc-new-template", line 10, in <module>
    sys.exit(new_template())
  File "/Users/leahwasser/opt/miniconda3/envs/earth-analytics-python/lib/python3.8/site-packages/abcclassroom/__main__.py", line 190, in new_template
    template.new_update_template(args)
  File "/Users/leahwasser/opt/miniconda3/envs/earth-analytics-python/lib/python3.8/site-packages/abcclassroom/template.py", line 59, in new_update_template
    token = cf.get_github_auth()["token"]
  File "/Users/leahwasser/opt/miniconda3/envs/earth-analytics-python/lib/python3.8/site-packages/ruamel/yaml/comments.py", line 753, in __getitem__
    return ordereddict.__getitem__(self, key)
KeyError: 'token'

```

lwasser commented:
note i'm in a new PR because i wanted to get this merged and had to do some cleanup. i wasn't sure how to push directly to your pr  without a new  clone.

lwasser commented:
the other  question i had... does the token authorize  access to all organizations that i have permissions to push to?  is that how it works?

lwasser commented:
i actually don't see the function `cf.get_github_auth()` unless i'm missing it? @kcranston did you change the config file? or am i missing maybe some changes  here? maybe you can  just add to  this  pr but either  i did  something wrong or  there is code missing or i 'm just missing something!

lwasser commented:
Ok progress -- the  function was an existing one. it returns this:

```
config["github"]
Out[46]: ordereddict([('access_token', 'a5eac40b6bc4d8f2faeef2e040c09fb190c3a5bd')])
```

There is no "token" key. do you just want to  access t hat token string for the template function? is that all we need to change perhaps?

kcranston commented:
> the other question i had... does the token authorize access to all organizations that i have permissions to push to? is that how it works?

There are two pieces. The abc-classroom-bot needs to be installed on any organization that you want to grant access, and then the token allows us to read / write repos on those organizations. This token is much more restrictive than the previous personal access token, which allowed abc-classroom to do _anything_ your account could do.

kcranston commented:
> Ok progress -- the function was an existing one. it returns this:
> 
> ```
> config["github"]
> Out[46]: ordereddict([('access_token', 'a5eac40b6bc4d8f2faeef2e040c09fb190c3a5bd')])
> ```
> 
> There is no "token" key. do you just want to access t hat token string for the template function? is that all we need to change perhaps?

You are right - the key should be `access_token`, not token. I changed the name to match the name that API returned, but missed changing this in `template.py`.

lwasser commented:
thanks @kcranston what's odd about that is i was playing with the ordered dict and that syntax object["access_token"] didn't allow me to grab the token value locally. that is why i didn't make the code change. i'll test this again locally. thank you!

lwasser commented:
ok still running into issues. it is not pushing files. will document here for the time being. @kcranston did this all work for you locally - did you test all of the functionality as it's not working for me still.
***
1.  minor - but no matter what i do it tells me it's changing the branch name to main. i need to see if that is just a print statement vs. it's trying to rename each time. it seems like that should only happen once

<img width="1199" alt="Screen Shot 2020-12-01 at 11 11 37 AM" src="https://user-images.githubusercontent.com/7649194/100779790-1e4f8080-33c6-11eb-8ca3-9e92d8d5a4f4.png">
2. when i try to push the template i get this:

<img width="1061" alt="Screen Shot 2020-12-01 at 11 13 43 AM" src="https://user-images.githubusercontent.com/7649194/100779926-58208700-33c6-11eb-9ee1-3056ad79eee3.png">
it creates a new repo but doesn't push the files. the odd thing is there is nothing in the repo to cause a conflict so i'm not sure why it's causing issues but it's not pushing successfully

lwasser commented on abcclassroom/github.py:
```diff
@@ -182,6 +307,27 @@ def init_and_commit(directory, custom_message=False):
+    the default branch to 'main', we need the local repo to match
+    in order to be able to push with error later.
+    """
+    print(
```
SO this runs each time even if the main branch exists. i think the thinking probably was for `new_template` you'd only run it once but i use it all of the time because the update template opens an annoying text editor. i think this would be more robust if it checked to see if main exists and renames accordingly. i could see this being problematic however if catching git returns is hard. 

This works but how hard is it to capture git output at the CLI using subprocess calls?
```
(base) CIRES-EL-LM-020:my-assignment-template leahwasser$ git rev-parse --verify main
070a6cc0cca92cb3dc034fe1832529eb54fc7afb
(base) CIRES-EL-LM-020:my-assignment-template leahwasser$ git rev-parse --verify mainwe
fatal: Needed a single revision
```

lwasser commented on abcclassroom/github.py:
```diff
@@ -182,6 +307,27 @@ def init_and_commit(directory, custom_message=False):
+    )
+    try:
+        _call_git("branch", "-m", "master", "main", directory=directory)
+    except RuntimeError:
```
@kcranston this doesn't seem to raise a Runtime error. Here is what happens. Can we capture the stdout (what is the correct type of output here) and check to see if it's fatal vs
```
# When main exists
(base) CIRES-EL-LM-020:my-assignment-template leahwasser$ git rev-parse --verify main
070a6cc0cca92cb3dc034fe1832529eb54fc7afb
# When a branch doesnt' exist  this is returned
(base) CIRES-EL-LM-020:my-assignment-template leahwasser$ git rev-parse --verify mainwe
fatal: Needed a single revision
(base) CIRES-EL-LM-020:my-assignment-template leahwasser$ git branch -m master main
fatal: A branch named 'main' already exists.
```

jedbrown replied:
I'd recommend `git show-ref --quiet --verify refs/heads/master`, which prints nothing and returns zero if the branch `master` exists, but not if a tag called `master` exists.

lwasser replied:
thank you @jedbrown !! we will look into this!!

lwasser commented:
This pr is LOADED with bugs.
I got the update-template to work - easy fix just switched master --> main but now i tried update template.

1. it's still opening VIM even tho my default editor is atom (not sure how that works when git default editor is already setup
2. it's returning the error below. maybe there are more masters somewhere??  

<img width="1301" alt="Screen Shot 2020-12-01 at 1 38 42 PM" src="https://user-images.githubusercontent.com/7649194/100794278-90ca5b80-33da-11eb-97d7-8fcce4ee6bfb.png">

kcranston commented:
The git editor is a separate outstanding issue - #311 . We should check for both the environment variable and the value of git's `core.editor` (these are separate variables).

kcranston commented:
I'll look into the other issues!

lwasser commented:
thank you @kcranston !! i think i fixed a bunch of issues. 
i'm not working on this PR at all i just fixed a few docs issues in a separate PR. i think what's left is that issue of it always trying to change master --> main and there is no true catch if this fails i think? 

i can't recreate that last screen shot that i posted again. i'm not sure if we should worry about it for now given i can't recreate it. in general template is working for me locally now but i would like to get rid of that always changing the branch name issue and have it do a check. in the future we won't need this if git does follow github (i hope they do).

kcranston commented:
Merging this, as @lwasser has been using this branch locally for some time and new authentication is working. See #363 and #311 for issues noted here that are not fixed in this PR.
